src/VisualStudio/TestUtilities/Remote/InProcRemostHostClient.cs

```diff
@@ -0,0 +1,220 @@
+
+            // get stream from service hub to communicate service specific information
+            // this is what consumer actually use to communicate information
+            var serviceStream = await _inprocServices.RequestServiceAsync(serviceName, cancellationToken).ConfigureAwait(false);
+
+            return new JsonRpcSession(snapshot, snapshotStream, callbackTarget, serviceStream, cancellationToken);
+        }
+
+        protected override void OnConnected()
+        {
+        }
+
+        protected override void OnDisconnected()
+        {
+            _rpc.Dispose();
+            _stream.Dispose();
+        }
+
+        private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
+        {
+            Disconnected();
+        }
+
+        private class InProcRemoteServices
+        {
+            private static readonly IServiceProvider s_serviceProvider = new ServiceProvider();
+
+            public InProcRemoteServices()
+            {
+            }
+
+            public Task<Stream> RequestServiceAsync(string serviceName, CancellationToken cancellationToken)
+            {
+                switch (serviceName)
+                {
+                    case WellKnownRemoteHostServices.RemoteHostService:
+                        {
+                            var tuple = FullDuplexStream.CreateStreams();
+                            return Task.FromResult<Stream>(new WrappedStream(new RemoteHostService(tuple.Item1, s_serviceProvider), tuple.Item2));
+                        }
+                    case hub::Microsoft.CodeAnalysis.Remote.WellKnownServiceHubServices.CodeAnalysisService:
+                        {
+                            var tuple = FullDuplexStream.CreateStreams();
+                            return Task.FromResult<Stream>(new WrappedStream(new CodeAnalysisService(tuple.Item1, s_serviceProvider), tuple.Item2));
+                        }
+                    case hub::Microsoft.CodeAnalysis.Remote.WellKnownServiceHubServices.SnapshotService:
+                        {
+                            var tuple = FullDuplexStream.CreateStreams();
+                            return Task.FromResult<Stream>(new WrappedStream(new SnapshotService(tuple.Item1, s_serviceProvider), tuple.Item2));
+                        }
+                }
+
+                throw ExceptionUtilities.UnexpectedValue(serviceName);
+            }
+
+            private class ServiceProvider : IServiceProvider
+            {
+                private static readonly TraceSource s_traceSource = new TraceSource("inprocRemoteClient");
+
+                public object GetService(Type serviceType)
+                {
+                    if (typeof(TraceSource) == serviceType)
+                    {
+                        return s_traceSource;
+                    }
+
+                    throw ExceptionUtilities.UnexpectedValue(serviceType);
+                }
+            }
+
+            private class WrappedStream : Stream
```

Comment:
when stream is closed, we need to dispose service as well. wrapped stream will tie two object's lifetime together